packages/appium/lib/config-file.js

```diff
@@ -0,0 +1,389 @@
+      '.yml': yamlLoader,
+      '.json': jsonLoader,
+      noExt: jsonLoader,
+    },
+  });
+  return (
+    (await (filepath ? loadConfigFile(lc, filepath) : searchConfigFile(lc))) ??
+    {}
+  );
+}
+
+/**
+ * Given an object, validates it against the Appium config schema.
+ * If errors occur, the returned array will be non-empty.
+ * @todo Provide API to validate against driver/plugin schemas.
+ * @public
+ * @param {any} value - The value (hopefully an object) to validate against the schema
+ * @returns {import('ajv').ErrorObject[]} Array of errors, if any.
+ */
+export function validateConfig (value) {
+  const validator = getValidator(APPIUM_CONFIG_SCHEMA_ID);
+  if (!validator) {
+    throw new Error(
+      `Could not find schema with ID: ${APPIUM_CONFIG_SCHEMA_ID}`,
+    );
+  }
+  return !validator(value) &&
+    _.isArray(validator.errors)
+    ? [...validator.errors]
+    : [];
+}
+
+/**
+ * Given an optional path, read a config file.
+ * @param {string} [filepath] - Path to config file, if we have one
+ * @param {ReadConfigFileOptions} [opts] - Options
+ * @returns {Promise<ReadConfigFileResult>} Contains config and filepath, if found, and any errors
+ */
+export async function readConfigFile (filepath, opts = {}) {
+  const result = await findConfigFile(filepath);
+
+  if (result.config && result.filepath && !result.isEmpty) {
+    log.debug(`Config file found at ${result.filepath}`);
+    const {normalize = true, pretty = true} = opts;
+    try {
+      /** @type {ReadConfigFileResult} */
+      let configResult;
+      const errors = validateConfig(result.config);
+
+      if (_.isEmpty(errors)) {
+        configResult = {...result, errors};
+      } else {
+        const reason = formatErrors(errors, result, {
+          json: rawConfig.get(result.filepath),
+          pretty,
+        });
+        configResult = reason
+          ? {...result, errors, reason}
+          : {...result, errors};
+      }
+
+      if (normalize) {
+        // normalize (to camel case) all top-level property names of the config file
+        configResult.config = normalizeConfig(
+          /** @type {AppiumConfiguration} */ (configResult.config),
+        );
+      }
+
+      log.verbose(format('Final config result: %O', configResult));
+
+      return configResult;
+    } finally {
+      // clean up the raw config file cache, which is only kept to better report errors.
+      rawConfig.delete(result.filepath);
+    }
+  }
+  return result;
+}
+
+/**
+ * Convert schema property names to either a) the value of the `appiumDest` property, if any; or b) camel-case
+ * @param {AppiumConfiguration} config - Configuration object
+ * @returns {NormalizedAppiumConfiguration} New object with camel-cased keys.
+ */
+function normalizeConfig (config) {
+  const jsonSchema =
+    /** @type {import('./schema').AppiumConfigJsonSchemaType} */ (
+      getValidator(APPIUM_CONFIG_SCHEMA_ID)?.schema
+    );
+  if (jsonSchema) {
+    return _.mapKeys(
```

Comment:
since it's a larger block of code, can we separate out rather than nest the mapping functions?

Reply:
there's already enough logical duplication in this set of maps, should we just go full recursive at this point?

Reply:
yeah, I don't like this function.  working on something better